File: backend/src/routes/leaderboard.js
const express = require('express');
const db = require('../config/database');
const { authenticate, optionalAuth } = require('../middleware/auth');
const logger = require('../config/logger');
const { getLeaderboardColumns } = require('../services/leaderboardColumns');

const router = express.Router();

// Get current period keys
const getPeriodKey = (period) => {
  const now = new Date();
  if (period === 'weekly') {
    const startOfYear = new Date(now.getFullYear(), 0, 1);
    const week = Math.ceil(((now - startOfYear) / 86400000 + startOfYear.getDay() + 1) / 7);
    return `${now.getFullYear()}-W${String(week).padStart(2, '0')}`;
  }
  if (period === 'monthly') return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}`;
  return 'all';
};

// ─── GET /api/leaderboard ─────────────────────────────────────────────────────
// Query: ?scope=global|country|school|class&period=weekly|monthly|all_time&limit=50
router.get('/', optionalAuth, async (req, res) => {
  const { scope = 'global', period = 'weekly', limit = 50, scopeId } = req.query;
  const periodKey = getPeriodKey(period);
  const user = req.user;

  try {
    const cols = await getLeaderboardColumns();
    const redis = req.app?.locals?.redis;
    const scopeKey = scopeId || user?.country || 'all';
    const cacheKey = `lb:${scope}:${period}:${periodKey}:${scopeKey}:${limit}:${user?.id || 'anon'}`;

    if (redis) {
      const cached = await redis.get(cacheKey);
      if (cached) return res.json(JSON.parse(cached));
    }

    const params = [user?.id || null, scope, period, periodKey];
    let paramIndex = params.length + 1;
    const filters = [];

    if (scope === 'country') {
      const country = scopeId || user?.country;
      if (country && cols.country) {
        filters.push(`AND le.country = $${paramIndex}`);
        params.push(country);
        paramIndex += 1;
      } else if (country && !cols.country) {
        logger.warn('leaderboard_entries.country missing; skipping country filter');
      }
    } else if (scope === 'school' || scope === 'class') {
      if (scopeId && cols.scope_id) {
        filters.push(`AND le.scope_id = $${paramIndex}`);
        params.push(scopeId);
        paramIndex += 1;
      } else if (scopeId && !cols.scope_id) {
        logger.warn('leaderboard_entries.scope_id missing; skipping scope filter');
      }
    }

    const limitParam = `$${paramIndex}`;
    params.push(parseInt(limit));

    const rankCol = cols.rank ? 'le.rank' : 'NULL::int as rank';
    const streakCol = cols.streak ? 'le.streak' : '0::int as streak';
    const testsCol = cols.tests_taken ? 'le.tests_taken' : '0::int as tests_taken';
    const avgCol = cols.avg_score ? 'le.avg_score' : '0::numeric as avg_score';

    const query = `
      SELECT 
        ${rankCol},
        le.xp,
        ${streakCol},
        ${testsCol},
        ${avgCol},
        u.id,
        u.name,
        u.avatar_url,
        u.country,
        u.grade_level,
        u.school_id,
        s.name AS school_name,
        CASE WHEN u.id = $1 THEN true ELSE false END AS is_current_user
      FROM leaderboard_entries le
      JOIN users u ON u.id = le.user_id
      LEFT JOIN schools s ON s.id = u.school_id
      WHERE le.scope = $2 
        AND le.period = $3 
        AND le.period_key = $4
        ${filters.join('\n')}
      ORDER BY le.xp DESC, le.streak DESC
      LIMIT ${limitParam}
    `;

    const { rows } = await db.query(query, params);

    // Get current user's rank if authenticated
    let userRank = null;
    if (user) {
      const userRankCols = [
        cols.rank ? 'rank' : 'NULL::int as rank',
        'xp',
        cols.streak ? 'streak' : '0::int as streak',
      ].join(', ');
      const rankResult = await db.query(
        `SELECT ${userRankCols} FROM leaderboard_entries 
         WHERE user_id = $1 AND scope = $2 AND period = $3 AND period_key = $4`,
        [user.id, scope, period, periodKey]
      );
      userRank = rankResult.rows[0] || null;
    }

    const payload = {
      leaderboard: rows,
      userRank,
      scope,
      period,
      periodKey,
      totalEntries: rows.length,
    };

    if (redis) await redis.set(cacheKey, JSON.stringify(payload), { EX: 60 });
    res.json(payload);

  } catch (err) {
    logger.error('Leaderboard fetch error', { error: err.message, stack: err.stack });
    res.status(500).json({ error: 'Failed to fetch leaderboard' });
  }
});

// ─── GET /api/leaderboard/my-ranks ───────────────────────────────────────────
router.get('/my-ranks', authenticate, async (req, res) => {
  const user = req.user;
  const periods = ['weekly', 'monthly', 'all_time'];
  const scopes = ['global', 'country', 'school'];
  const results = {};

  try {
    for (const period of periods) {
      results[period] = {};
      const periodKey = getPeriodKey(period);
      for (const scope of scopes) {
        const { rows } = await db.query(
          `SELECT rank, xp, streak, tests_taken, avg_score FROM leaderboard_entries
           WHERE user_id = $1 AND scope = $2 AND period = $3 AND period_key = $4`,
          [user.id, scope, period, periodKey]
        );
        results[period][scope] = rows[0] || { rank: null, xp: 0 };
      }
    }
    res.json({ ranks: results });
  } catch (err) {
    logger.error('My ranks error:', err.message);
    res.status(500).json({ error: 'Failed to fetch your rankings' });
  }
});

module.exports = router;

File: backend-py/services/learning-service/app/routers/leaderboard.py
"""Leaderboard routes (mirrors backend/src/routes/leaderboard.js)."""
from __future__ import annotations

import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from ..deps import AdminOnly, current_principal, get_principal, get_session

router = APIRouter(prefix="/api/leaderboard", tags=["leaderboard"])


def _period_key(period: str) -> str:
    now = datetime.utcnow()
    if period == "weekly":
        year, week, _ = now.isocalendar()
        return f"{year}-W{week:02d}"
    if period == "monthly":
        return f"{now.year}-{now.month:02d}"
    return "all"


@router.get("")
async def leaderboard(
    scope: str = Query("global"),
    period: str = Query("weekly"),
    limit: int = Query(50, ge=1, le=200),
    scope_id: uuid.UUID | None = Query(None, alias="scopeId"),
    principal=current_principal,
    sess: AsyncSession = Depends(get_session),
):
    period_key = _period_key(period)
    user_id = uuid.UUID(principal.user_id) if principal else None

    filters = []
    params = {"uid": user_id, "sc": scope, "p": period, "pk": period_key, "lim": limit}
    if scope == "country":
        # Country needs looking up on the user
        row = (await sess.execute(
            text("SELECT country FROM users WHERE id = :uid"), {"uid": user_id}
        )).mappings().first()
        country = (row and row["country"]) if row else None
        if country:
            filters.append("AND le.country = :ctry")
            params["ctry"] = country
    elif scope == "class":
        # "Class" leaderboard = students at the same grade_level. If the caller belongs
        # to a school, we further restrict to that school so it reads as *their* class,
        # not every student globally at that grade.
        row = (await sess.execute(
            text("SELECT grade_level, school_id, country FROM users WHERE id = :uid"),
            {"uid": user_id},
        )).mappings().first()
        grade_level = row and row["grade_level"]
        caller_school = row and row["school_id"]
        caller_country = row and row["country"]
        if not grade_level:
            # No class set yet — return an empty list rather than a 500.
            return {
                "leaderboard": [],
                "userRank": None,
                "scope": scope,
                "period": period,
                "periodKey": period_key,
                "totalEntries": 0,
                "note": "no_grade_level",
            }
        # Query on global scope entries because we don't materialise per-class entries;
        # instead we filter users by grade_level.
        params["sc"] = "global"
        filters.append("AND u.grade_level = :grade")
        params["grade"] = grade_level
        if caller_school:
            filters.append("AND u.school_id = :school")
            params["school"] = caller_school
        elif caller_country:
            filters.append("AND u.country = :ctry")
            params["ctry"] = caller_country
        # Only compare students to other students.
        filters.append("AND u.role = 'student'")
    elif scope in ("school", "class") and scope_id:
        filters.append("AND le.scope_id = :sid")
        params["sid"] = scope_id

    # The public leaderboard is a student-only ranking across all scopes.
    # Teachers/admins are ranked separately via /api/leaderboard/school-rankings.
    if "AND u.role = 'student'" not in filters:
        filters.append("AND u.role = 'student'")

    q = text(
        f"""
        SELECT
          le.rank, le.xp, le.streak, le.tests_taken, le.avg_score,
          u.id, u.name, u.avatar_url, u.country, u.grade_level, u.school_id,
          s.name AS school_name,
          CASE WHEN u.id = :uid THEN TRUE ELSE FALSE END AS is_current_user
        FROM leaderboard_entries le
        JOIN users u ON u.id = le.user_id
        LEFT JOIN schools s ON s.id = u.school_id
        WHERE le.scope = :sc AND le.period = :p AND le.period_key = :pk
        {' '.join(filters)}
        ORDER BY le.xp DESC, le.streak DESC
        LIMIT :lim
        """
    )
    rows = (await sess.execute(q, params)).mappings().all()

    user_rank = None
    if user_id:
        rank_row = (await sess.execute(
            text(
                "SELECT rank, xp, streak FROM leaderboard_entries le "
                "JOIN users u ON u.id = le.user_id "
                "WHERE le.user_id = :uid AND le.scope = :sc "
                "AND le.period = :p AND le.period_key = :pk "
                "AND u.role = 'student'"
            ),
            {"uid": user_id, "sc": params["sc"], "p": period, "pk": period_key},
        )).mappings().first()
        if rank_row:
            user_rank = dict(rank_row)

    return {
        "leaderboard": [dict(r) for r in rows],
        "userRank": user_rank,
        "scope": scope,
        "period": period,
        "periodKey": period_key,
        "totalEntries": len(rows),
    }


@router.get("/my-ranks")
async def my_ranks(
    principal=current_principal,
    sess: AsyncSession = Depends(get_session),
):
    uid = uuid.UUID(principal.user_id)
    out: dict[str, dict[str, dict]] = {}
    for period in ("weekly", "monthly", "all_time"):
        out[period] = {}
        pk = _period_key(period)
        for scope in ("global", "country", "school"):
            row = (await sess.execute(
                text(
                    "SELECT rank, xp, streak, tests_taken, avg_score FROM leaderboard_entries "
                    "WHERE user_id = :uid AND scope = :sc AND period = :p AND period_key = :pk"
                ),
                {"uid": uid, "sc": scope, "p": period, "pk": pk},
            )).mappings().first()
            out[period][scope] = dict(row) if row else {"rank": None, "xp": 0}
    return {"ranks": out}


@router.get("/school-rankings")
async def school_rankings(
    period: str = Query("weekly"),
    principal=AdminOnly,
    sess: AsyncSession = Depends(get_session),
):
    """School admin: student + teacher-aggregated rankings."""
    # Find requester's school
    me_row = (await sess.execute(
        text("SELECT school_id FROM users WHERE id = :uid"),
        {"uid": uuid.UUID(principal.user_id)},
    )).mappings().first()
    school_id = me_row and me_row["school_id"]
    if not school_id:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No school associated")

    period_key = _period_key(period)

    student_rankings = (await sess.execute(
        text(
            """
            SELECT le.xp, le.streak, le.tests_taken, le.avg_score,
                   u.id, u.name, u.grade_level, u.avatar_url
            FROM leaderboard_entries le
            JOIN users u ON u.id = le.user_id
            WHERE u.school_id = :sid AND u.role = 'student'
              AND le.scope = 'school' AND le.period = :p AND le.period_key = :pk
            ORDER BY le.xp DESC LIMIT 100
            """
        ),
        {"sid": school_id, "p": period, "pk": period_key},
    )).mappings().all()

    teacher_rankings = (await sess.execute(
        text(
            """
            SELECT t.id AS teacher_id, t.name AS teacher_name,
                   COUNT(DISTINCT le.user_id) AS student_count,
                   COALESCE(SUM(le.xp), 0) AS total_xp,
                   COALESCE(ROUND(AVG(le.xp)), 0) AS avg_xp,
                   COALESCE(ROUND(AVG(le.avg_score), 1), 0) AS avg_score,
                   COALESCE(SUM(le.tests_taken), 0) AS total_tests,
                   COALESCE(ROUND(AVG(le.streak), 1), 0) AS avg_streak
            FROM users t
            LEFT JOIN users s ON s.school_id = t.school_id AND s.role = 'student'
            LEFT JOIN leaderboard_entries le ON le.user_id = s.id
              AND le.scope = 'school' AND le.period = :p AND le.period_key = :pk
            WHERE t.school_id = :sid AND t.role = 'teacher'
            GROUP BY t.id, t.name
            ORDER BY avg_xp DESC
            """
        ),
        {"sid": school_id, "p": period, "pk": period_key},
    )).mappings().all()

    return {
        "studentRankings": [dict(r) for r in student_rankings],
        "teacherRankings": [dict(r) for r in teacher_rankings],
        "period": period,
        "periodKey": period_key,
    }
